src/main/java/org/opentripplanner/model/transfer/TransferService.java

```diff
@@ -1,38 +1,52 @@
 package org.opentripplanner.model.transfer;
 
-import com.google.common.collect.ArrayListMultimap;
-import com.google.common.collect.Multimap;
 import java.io.Serializable;
 import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
 import org.opentripplanner.common.model.P2;
+import org.opentripplanner.common.model.T2;
 import org.opentripplanner.model.Stop;
 import org.opentripplanner.model.Trip;
-
-// TODO OTP2 reimplement all special kinds of transfers
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * This class represents all transfer information in the graph. Transfers are grouped by
  * stop-to-stop pairs.
+ * <p>
+ * THIS CLASS IS NOT THREAD_SAFE. This class is loaded with plan data only, and read-only during
+ * routing. No real-time update should touch this class; Hence it do not need to be thread-safe.
  */
 public class TransferService implements Serializable {
 
+    private static final Logger LOG = LoggerFactory.getLogger(TransferService.class);
+
     /**
-     * Table which contains transfers between two stops
+     * Table which contains transfers between two trips/routes
      */
-    protected Multimap<P2<Stop>, Transfer> table = ArrayListMultimap.create();
+    private final Map<P2<TripTransferPoint>, Transfer> trip2tripTransfers;
 
-    private Transfer getTransfer(Stop fromStop, Stop toStop, Trip fromTrip, Trip toTrip) {
-        Collection<Transfer> transfers = table.get(new P2<>(fromStop, toStop));
-        for (Transfer transfer : transfers) {
-            if (transfer.getFromTrip() == fromTrip && transfer.getToTrip() == toTrip) {
-                return transfer;
-            }
-        }
-        return null;
-    }
+    /**
+     * Table which contains transfers between a trip/route and a stops
+     */
+    private final Map<T2<TripTransferPoint, Stop>, Transfer> trip2StopTransfers;
 
-    public Collection<Transfer> getTransfers() {
-        return table.values();
+    /**
+     * Table which contains transfers between a stop and a trip/route
+     */
+    private final Map<T2<Stop, TripTransferPoint>, Transfer> stop2TripTransfers;
+
+    /**
+     * Table which contains transfers between two stops
+     */
+    private final Map<P2<Stop>, Transfer> stop2StopTransfers;
+
+    public TransferService() {
+        this.trip2tripTransfers = new HashMap<>();
```

Comment:
Capitalization of second word is inconsistent with identifiers below. Use of phonetic 2 for "To" only saves one character and may be slightly more confusing.